### PR TITLE
Skip evaluation for rows with errors in Expr::evalSimplified

### PR DESCRIPTION
**Pull request by kgpai**
This PR is an unabashed reproduction of @kevinwilfong's PR (https://github.com/facebookincubator/velox/pull/1455) with very slight modifications. 

We need to deselect rows with errors since otherwise they will be cleared by Conjunct special form , as current assumption is that they are only invoked on rows that do not have exceptions.

For e.g : try(regexp_extract(json_extract_scalar(X,"c0"),"c1")) , will still run regexp_extract for rows where json_extract_scalar has thrown. 
In this PR we filter out those rows that have thrown exceptions.

### CODE REVIEW

File: velox/expression/Expr.cpp [hunk 1]
```diff
@@ -312,6 +312,20 @@ void Expr::evalSimplifiedImpl(
     }
   }
 
+  // If any errors occurred evaluating the arguments, its possible that the
+  // values for those arguments were not defined which can lead to undefined
+  // behavior if we try and evauate current function on them. The value for this
```
Comment (mbasmanova) [inline velox/expression/Expr.cpp:317]: typo: evauate -> evaluate

File: velox/expression/tests/TryExprTest.cpp [hunk 2]
```diff
@@ -130,6 +130,37 @@ TEST_F(TryExprTest, nestedTryParentErrors) {
...
 }
 
+TEST_F(TryExprTest, skipExecutionEvalSimplified) {
+  registerFunction<CountCallsFunction, int64_t, int64_t>(
+      {"count_calls"}, BIGINT());
+
+  // Test that when a subset of the inputs to a function wrapped in a TRY throw
+  // exceptions, that function is only evaluated on the inputs that did not
+  // throw exceptions.
+  std::vector<std::optional<int64_t>> expected{
```
Comment (mbasmanova) [inline velox/expression/tests/TryExprTest.cpp:140]: nit: for readability, consider turning expected into a Velox vector and moving it just before it is used:

```
auto expected = makeNullableFlatVector<int64_t>({0, std::nullopt, 1, std::nullopt, 2});
assertEqualVectors(expected, result);
```

File: velox/expression/tests/TryExprTest.cpp [hunk 3]
```diff
@@ -130,6 +130,37 @@ TEST_F(TryExprTest, nestedTryParentErrors) {
...
+
+  // Test that when all the inputs to a function wrapped in a TRY throw
+  // exceptions, that function isn't evaluated and a NULL constant is returned
+  // directly.
+  std::vector<std::optional<int64_t>> expected(3, std::nullopt);
+  auto flatVector = makeFlatVector<StringView>({"a", "b", "c"});
+  auto result = evaluateSimplified<ConstantVector<int64_t>>(
+      "try(count_calls(cast(c0 as integer)))", makeRowVector({flatVector}));
+
+  assertEqualVectors(makeNullableFlatVector(expected), result);
```
Comment (mbasmanova) [inline velox/expression/tests/TryExprTest.cpp:161]: nit: 

```
auto expected = makeNullConstant(TypeKind::BIGINT, 3);
```

**Review by mbasmanova — APPROVED** [review]
@kgpai The change looks good to me, but I wonder what was the problem without that change. The values in the rows with error may be garbage and this may cause the function to generate an error for these, but TRY should convert all these errors to NULL at the end, hence, I think the result should be the same. 

CC: @bikramSingh91

**Comment by kgpai** [comment]
Hi @mbasmanova , The major reason I think we want to do this is we do not want function evaluation on rows that have garbage values , the end result as you note might be same, however its probably safer to not allow this to happen (perhaps functions have side effects, or request too much memory etc..).

**Comment by mbasmanova** [comment]
> its probably safer to not allow this to happen (perhaps functions have side effects, or request too much memory etc..).

Functions are not supposed to have side effects or request too much memory. In a way, it would be useful to catch these cases and fix the functions.

**Comment by kgpai** [comment]
Abandoning this PR , since we would like to expose bugs in udfs.

File: velox/expression/Expr.cpp [hunk 4]
```diff
@@ -312,6 +312,21 @@ void Expr::evalSimplifiedImpl(
     }
   }
 
+  // If any errors occurred evaluating the arguments, its possible that the
+  // values for those arguments were not defined which can lead to undefined
```
Comment (mbasmanova) [inline velox/expression/Expr.cpp:316]: @kgpai @oerling Krishna, turns out we do need this change, but for a different reason. Conjunct special form (AND/OR) assumes that it is invoked only on rows that do not have errors yet. If there are any pre-existing errors on 'rows' passed to Conjunct special form, these will be cleared (unless new errors occur during evaluation of the conjunct inputs). See finalizeErrors in velox/expression/ConjunctExpr.cpp.

**Pull request closed without merging**